### PR TITLE
Add resize container and exec command

### PR DESCRIPTION
**Pull request by creaink**
Improve resize exec test and adapt project struct for #1106

---
This change is [Reviewable](https://reviewable.io/reviews/docker-java/docker-java/1262)

### CODE REVIEW

File: docker-java/src/test/java/com/github/dockerjava/cmd/ResizeExecCmdIT.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        assertThat(container.getId(), not(isEmptyString()));
+
+        dockerRule.getClient().startContainerCmd(container.getId()).exec();
+
+        ExecCreateCmdResponse execCreateCmdResponse = dockerRule.getClient().execCreateCmd(container.getId()).withTty(true)
+                 .withAttachStdout(true).withAttachStderr(true).withCmd("sh").exec();
+
+        final ExecStartResultCallback execStartResultCallback = new ExecStartResultCallback(System.out, System.err);
+
+        Thread execThread = new Thread(new Runnable() {
```
Comment (bsideup) [inline docker-java/src/test/java/com/github/dockerjava/cmd/ResizeExecCmdIT.java:38]: this thread is not needed

File: docker-java-transport-jersey/src/main/java/com/github/dockerjava/jaxrs/ResizeExecCmdExec.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ResizeExecCmdExec.class);
+
+    public ResizeExecCmdExec(WebTarget baseResource, DockerClientConfig dockerClientConfig) {
+        super(baseResource, dockerClientConfig);
+    }
+
+    @Override
+    protected Void execute(ResizeExecCmd command) {
+        WebTarget webResource = getBaseResource().path("/exec/{id}/resize").resolveTemplate("id",
```
Comment (bsideup) [inline docker-java-transport-jersey/src/main/java/com/github/dockerjava/jaxrs/ResizeExecCmdExec.java:21]: please split into method-by-line

File: docker-java-transport-jersey/src/main/java/com/github/dockerjava/jaxrs/ResizeContainerCmdExec.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ResizeContainerCmdExec.class);
+
+    public ResizeContainerCmdExec(WebTarget baseResource, DockerClientConfig dockerClientConfig) {
+        super(baseResource, dockerClientConfig);
+    }
+
+    @Override
+    protected Void execute(ResizeContainerCmd command) {
+        WebTarget webResource = getBaseResource().path("/containers/{id}/resize")
```
Comment (bsideup) [inline docker-java-transport-jersey/src/main/java/com/github/dockerjava/jaxrs/ResizeContainerCmdExec.java:21]: see https://github.com/docker-java/docker-java/pull/1262/files#r432382962

File: docker-java-core/src/main/java/com/github/dockerjava/core/exec/ResizeExecCmdExec.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ResizeExecCmdExec.class);
+
+    public ResizeExecCmdExec(WebTarget baseResource, DockerClientConfig dockerClientConfig) {
+        super(baseResource, dockerClientConfig);
+    }
+
+    @Override
+    protected Void execute(ResizeExecCmd command) {
+        WebTarget webResource = getBaseResource().path("/exec/{id}/resize").resolveTemplate("id",
```
Comment (bsideup) [inline docker-java-core/src/main/java/com/github/dockerjava/core/exec/ResizeExecCmdExec.java:21]: see https://github.com/docker-java/docker-java/pull/1262/files#r432382962

File: docker-java-core/src/main/java/com/github/dockerjava/core/exec/ResizeContainerCmdExec.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ResizeContainerCmdExec.class);
+
+    public ResizeContainerCmdExec(WebTarget baseResource, DockerClientConfig dockerClientConfig) {
+        super(baseResource, dockerClientConfig);
+    }
+
+    @Override
+    protected Void execute(ResizeContainerCmd command) {
+        WebTarget webResource = getBaseResource().path("/containers/{id}/resize")
```
Comment (bsideup) [inline docker-java-core/src/main/java/com/github/dockerjava/core/exec/ResizeContainerCmdExec.java:20]: see https://github.com/docker-java/docker-java/pull/1262/files#r432382962

File: docker-java-core/src/main/java/com/github/dockerjava/core/command/ResizeExecCmdImpl.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    @Override
+    public ResizeExecCmd withSize(int height, int width) {
+        this.height = height;
+        this.width = width;
+        return this;
+    }
+
+    /**
+     * @throws NotFoundException
+     *             No such exec instance
```
Comment (bsideup) [inline docker-java-core/src/main/java/com/github/dockerjava/core/command/ResizeExecCmdImpl.java:55]:
```suggestion
     * @throws NotFoundException no such exec instance
```

File: docker-java-core/src/main/java/com/github/dockerjava/core/command/ResizeExecCmdImpl.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
+package com.github.dockerjava.core.command;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonInclude.Include;
+import com.github.dockerjava.api.command.ResizeExecCmd;
+import com.github.dockerjava.api.exception.NotFoundException;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+@JsonInclude(Include.NON_NULL)
```
Comment (bsideup) [inline docker-java-core/src/main/java/com/github/dockerjava/core/command/ResizeExecCmdImpl.java:10]: this class isn't used as a JSON object
```suggestion
```

File: docker-java-core/src/main/java/com/github/dockerjava/core/command/ResizeContainerCmdImpl.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    @Override
+    public ResizeContainerCmd withSize(int height, int width) {
+        this.height = height;
+        this.width = width;
+        return this;
+    }
+
+    /**
+     * @throws NotFoundException
+     *             No such exec instance
```
Comment (bsideup) [inline docker-java-core/src/main/java/com/github/dockerjava/core/command/ResizeContainerCmdImpl.java:55]:
```suggestion
     * @throws NotFoundException no such exec instance
```

File: docker-java-core/src/main/java/com/github/dockerjava/core/command/ResizeContainerCmdImpl.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
+package com.github.dockerjava.core.command;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonInclude.Include;
+import com.github.dockerjava.api.command.ResizeContainerCmd;
+import com.github.dockerjava.api.exception.NotFoundException;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+@JsonInclude(Include.NON_NULL)
```
Comment (bsideup) [inline docker-java-core/src/main/java/com/github/dockerjava/core/command/ResizeContainerCmdImpl.java:10]: this class isn't used as a JSON object
```suggestion
```

File: docker-java-api/src/main/java/com/github/dockerjava/api/command/ResizeExecCmd.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+    Integer getWidth();
+
+    ResizeExecCmd withExecId(@Nonnull String execId);
+
+    ResizeExecCmd withSize(int height, int width);
+
+    /**
+     * @throws NotFoundException
+     *             No such exec instance
```
Comment (bsideup) [inline docker-java-api/src/main/java/com/github/dockerjava/api/command/ResizeExecCmd.java:22]:
```suggestion
     * @throws NotFoundException no such exec instance
```

File: docker-java-api/src/main/java/com/github/dockerjava/api/command/ResizeContainerCmd.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+    Integer getWidth();
+
+    ResizeContainerCmd withContainerId(@Nonnull String execId);
+
+    ResizeContainerCmd withSize(int height, int width);
+
+    /**
+     * @throws NotFoundException
+     *             No such container instance
```
Comment (bsideup) [inline docker-java-api/src/main/java/com/github/dockerjava/api/command/ResizeContainerCmd.java:23]:
```suggestion
     * @throws NotFoundException no such container instance
```

**Comment by bsideup** [comment]
@creaink could you please sync with master?

File: docker-java/src/test/java/com/github/dockerjava/cmd/ResizeExecCmdIT.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        assertThat(container.getId(), not(isEmptyString()));
+
+        dockerRule.getClient().startContainerCmd(container.getId()).exec();
+
+        ExecCreateCmdResponse execCreateCmdResponse = dockerRule.getClient().execCreateCmd(container.getId()).withTty(true)
+                 .withAttachStdout(true).withAttachStderr(true).withCmd("top").exec();
+
+        final ExecStartResultCallback execStartResultCallback = new ExecStartResultCallback(System.out, System.err);
+
+        Thread execThread = new Thread(() -> {
```
Comment (bsideup) [inline docker-java/src/test/java/com/github/dockerjava/cmd/ResizeExecCmdIT.java:39]: Why this thread? There is no point in offloading `awaitCompletion()` to a separate thread. Could you please clarify what you wanted to achieve?

File: docker-java/src/test/java/com/github/dockerjava/cmd/ResizeContainerCmdIT.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+        CreateContainerResponse container = dockerRule.getClient().createContainerCmd(DEFAULT_IMAGE).withUser("root")
+                .withCmd("sh").withName(containerName).withTty(true).withStdinOpen(true).exec();
+
+        LOG.info("Created container {}", container.toString());
+
+        assertThat(container.getId(), not(isEmptyString()));
+
+        dockerRule.getClient().startContainerCmd(container.getId()).exec();
+        dockerRule.getClient().resizeContainerCmd(container.getId()).withSize(30, 120).exec();
```
Comment (bsideup) [inline docker-java/src/test/java/com/github/dockerjava/cmd/ResizeContainerCmdIT.java:29]: is there some way to test that we actually resize the terminal?

File: docker-java/src/test/java/com/github/dockerjava/cmd/ResizeExecCmdIT.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+                execStartResultCallback.awaitCompletion();
+            } catch (InterruptedException ignored) {
+            }
+        });
+
+        dockerRule.getClient().execStartCmd(execCreateCmdResponse.getId()).exec(execStartResultCallback);
+        execThread.start();
+        Thread.sleep(TimeUnit.SECONDS.toMillis(3));
+
+        dockerRule.getClient().resizeExecCmd(execCreateCmdResponse.getId()).withSize(30, 120).exec();
```
Comment (bsideup) [inline docker-java/src/test/java/com/github/dockerjava/cmd/ResizeExecCmdIT.java:49]: Same question:
https://github.com/docker-java/docker-java/pull/1262/files#r436027566

File: docker-java/src/test/java/com/github/dockerjava/cmd/ResizeExecCmdIT.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+        Thread execThread = new Thread(() -> {
+            try {
+                execStartResultCallback.awaitCompletion();
+            } catch (InterruptedException ignored) {
+            }
+        });
+
+        dockerRule.getClient().execStartCmd(execCreateCmdResponse.getId()).exec(execStartResultCallback);
+        execThread.start();
+        Thread.sleep(TimeUnit.SECONDS.toMillis(3));
```
Comment (bsideup) [inline docker-java/src/test/java/com/github/dockerjava/cmd/ResizeExecCmdIT.java:47]: please avoid static delays

File: docker-java/src/test/java/com/github/dockerjava/cmd/ResizeExecCmdIT.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    public void resizeExecTest() throws Exception {
+        String containerName = "generated_" + new SecureRandom().nextInt();
+
+        CreateContainerResponse container = dockerRule.getClient().createContainerCmd(DEFAULT_IMAGE).withUser("root")
+                .withCmd("sleep", "9999").withName(containerName).exec();
+
+        LOG.info("Created container {}", container.toString());
+
+        assertThat(container.getId(), not(isEmptyString()));
+
```
Comment (bsideup) [inline docker-java/src/test/java/com/github/dockerjava/cmd/ResizeExecCmdIT.java:30]:
```suggestion
```

File: docker-java/src/test/java/com/github/dockerjava/cmd/ResizeContainerCmdIT.java [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+    @Test
+    public void resizeExecTest() {
+        String containerName = "generated_" + new SecureRandom().nextInt();
+
+        CreateContainerResponse container = dockerRule.getClient().createContainerCmd(DEFAULT_IMAGE).withUser("root")
+                .withCmd("sh").withName(containerName).withTty(true).withStdinOpen(true).exec();
+
+        LOG.info("Created container {}", container.toString());
+
+        assertThat(container.getId(), not(isEmptyString()));
```
Comment (bsideup) [inline docker-java/src/test/java/com/github/dockerjava/cmd/ResizeContainerCmdIT.java:26]:
```suggestion
```

File: docker-java/src/test/java/com/github/dockerjava/cmd/ResizeExecCmdIT.java (hunk 12 above)
Comment (creaink) [reply docker-java/src/test/java/com/github/dockerjava/cmd/ResizeExecCmdIT.java]: remove thread

File: docker-java/src/test/java/com/github/dockerjava/cmd/ResizeContainerCmdIT.java (hunk 13 above)
Comment (creaink) [reply docker-java/src/test/java/com/github/dockerjava/cmd/ResizeContainerCmdIT.java]: add validation of resize

**Comment by bsideup** [comment]
@creaink merged! Thank you!

**Comment by creaink** [comment]
@bsideup when I run test case on my computer, even I set environment variable like DOCKER_HOST.But it still require docker-java.properties. It is a bug or something?

**Comment by bsideup** [comment]
@creaink please share more info. How does it require? Throws an error or something? Also, which OS are you using?

**Comment by creaink** [comment]
> @creaink please share more info. How does it require? Throws an error or something? Also, which OS are you using?

OK, I will submit a issue with more detail about this later.

**Pull request merged**